### PR TITLE
Fix naive proxy retrieval

### LINKED ISSUE #3942: cli - naive proxy retrieval approach used causing errors
There's a naive approach used for getting the default (internal) capsule/smart proxy in the setUp() fixtures for `HostUpdateTestCase` and `HostDeleteTestCase`, causing test errors on irregular basis:

https://github.com/SatelliteQE/robottelo/blob/master/tests/foreman/cli/test_host.py#L334
https://github.com/SatelliteQE/robottelo/blob/master/tests/foreman/cli/test_host.py#L387
`edit`: API tests included:
https://github.com/SatelliteQE/robottelo/blob/master/tests/foreman/api/test_organization.py#L354

This approach does not count on possibility of another capsule being in place already, which might be printed as a first one (index 0). The sorting does not happen by id but by label (or a name?).

So from time to time, with no cleanup in place, test gets satellite with some capsule already in place considering it to be the internal one.

```
2016-10-25 14:50:25 - robottelo.ssh - DEBUG - >>> LANG=en_US.UTF-8  hammer -v -u admin -p changeme --output=csv proxy list --per-page="10000"
2016-10-25 14:50:28 - robottelo.ssh - DEBUG - <<< stdout
Id,Name,URL,Features
15,0eS0XZRGUE,https://sat6.com:9470,"Pulp, TFTP, DNS, DHCP, Puppet, Puppet CA, Dynflow, Discovery, Openscap, SSH"
1,sat6.com,https://sat6.com:9090,"Pulp, TFTP, DNS, DHCP, Puppet, Puppet CA, Dynflow, Discovery, Openscap, SSH"
...
2016-10-25 14:50:33 - robottelo.ssh - DEBUG - >>> LANG=en_US.UTF-8  hammer -v -u admin -p changeme --output=csv host create (...) --puppet-proxy-id="15"
```

another concurrent test, responsible for the fake capsule might in the meantime remove the capsule and, ta-daa:

```
ERROR 2016-10-25 14:50:36 Exception] ERROR:  insert or update on table "hosts" violates foreign key constraint "hosts_puppet_proxy_id_fk"
DETAIL:  Key (puppet_proxy_id)=(15) is not present in table "smart_proxies".
```

The test fails.

proposed fix:
use id=1 for the internal proxy (as it is always 1) or do the additional filtering.

### PR DESCRIPTION
**Pull request by svtkachenko**
Fixes #3942
Test run results for 6.3
```python
% nosetests tests/foreman/cli/test_host.py:{HostCreateTestCase,HostDeleteTestCase,HostUpdateTestCase,HostParameterTestCase}           
..............................
----------------------------------------------------------------------
Ran 30 tests in 1684.901s
OK
```
```python
% nosetests tests/foreman/cli/test_hostgroup.py:HostGroupTestCase.test_positive_create_with_puppet_ca_proxy 
.                                                                                                      
----------------------------------------------------------------------
Ran 1 test in 12.701s
OK
```
```python
% nosetests tests/foreman/api/test_host.py:HostTestCase.{test_positive_create_with_puppet_proxy,test_positive_create_with_puppet_ca_proxy,test_positive_update_puppet_proxy,test_positive_update_puppet_ca_proxy}
....python
----------------------------------------------------------------------
Ran 4 tests in 75.776s
OK
```
```python
% nosetests tests/foreman/api/test_hostgroup.py:HostGroupTestCase.{test_positive_create_with_puppet_ca_proxy,test_positive_create_with_realm,test_positive_create_with_puppet_proxy,test_positive_create_with_content_source,test_positive_update_puppet_ca_proxy,test_positive_update_realm,test_positive_update_puppet_proxy,test_positive_update_content_source}      
.S..SSSS
----------------------------------------------------------------------
Ran 8 tests in 20.023s
OK (SKIP=5)
```
`tests/foreman/api/test_smartproxy.py:SmartProxyMissingAttrTestCase` skipped becaus of bug.
`tests/foreman/api/test_organization.py:OrganizationUpdateTestCase.test_positive_add_smart_proxy` fails because of bug in 6.3. However, test passed with the change on 6.2.
`tests/foreman/cli/test_hostgroup.py:HostGroupTestCase.test_positive_create_with_multiple_entities` fails because of bug in 6.3. However, test passed with the change on 6.2.
`tests/foreman/api/test_location.py:LocationTestCase` tests are failing because of bug in 6.3. However, tests passed with the change on 6.2.

### CODE REVIEW

**Comment by svtkachenko** [comment]
Added changes for another 2 tests.
```python
% nosetests tests/foreman/cli/test_subnet.py:SubnetTestCase.test_positive_list tests/foreman/cli/test_gpgkey.py:TestGPGKey.test_positive_create_with_default_org
..
----------------------------------------------------------------------
Ran 2 tests in 132.769s
OK
```

**Comment by svtkachenko** [comment]
In #3942 two ways were proposed: use proxy id or add additional filtering.
At first I've decided to use id but @abalakh said that not always `id` can be 1 for default entity. For example, Default Location can have `id` 2.
So in another commit I've added search by name that is default to satellite hostname. But it requires FQDN name specified in properties file.

**Comment by rplevka** [comment]
> not always id can be 1 for default entity

For (internal) capsule, the ID of the default one is always 1

**Comment by rplevka** [comment]
the background behind 2 being a default location id seems to be the fact that orgs and locs share the same db and there is always a Default Org with id 1 followed by Default location being 2:

```
foreman=# select * from taxonomies;
 id |         name         |     type     |         created_at         |         updated_at         |      ignore_types      | description |        label         | ancestry |        title         | katello_default 
----+----------------------+--------------+----------------------------+----------------------------+------------------------+-------------+----------------------+----------+----------------------+-----------------
  2 | Default Location     | Location     | 2016-11-08 05:11:40.894822 | 2016-11-08 05:11:40.894822 | ---                   +|             |                      |          | Default Location     | t
    |                      |              |                            |                            | - ProvisioningTemplate+|             |                      |          |                      | 
    |                      |              |                            |                            | - Hostgroup           +|             |                      |          |                      | 
    |                      |              |                            |                            |                        |             |                      |          |                      | 
  1 | Default Organization | Organization | 2016-11-08 05:11:40.77291  | 2016-11-08 05:11:43.221976 | --- []                +|             | Default_Organization |          | Default Organization | f
    |                      |              |                            |                            |                        |             |                      |          |                      | 
  3 | rplevka_loc          | Location     | 2016-11-15 17:08:27.031495 | 2016-11-15 17:08:27.031495 | ---                   +|             |                      |          | rplevka_loc          | f
    |                      |              |                            |                            | - ProvisioningTemplate+|             |                      |          |                      | 
    |                      |              |                            |                            | - Hostgroup           +|             |                      |          |                      | 
    |                      |              |                            |                            |                        |             |                      |          |                      | 
(3 rows)
```

**Comment by abalakh** [comment]
@rplevka better not to rely on id=1 imo. We had numerous issues with lots of 'default' entities not present after roll back on jenkins (not sure which approach we currently use, foreman reset? or maybe vm snapshot?) and as a result entity with id 1 was not the default one, but some custom one created by test. 
I wanted to give you default domain as an example - as it wasn't created after roll back some time ago, but i've just ran one test on our qe machine and was surprised it failed. Failure looks like this:
```
hammer> medium create --operatingsystem-ids="1" --name="bJEclJ" --organization-ids="1" --location-ids="2" --path="http://VvRKBH"
Could not create the installation medium:
  Resource medium not found by id ''
```
Pretty weird message, yeah? The root cause is this:
```
hammer> os info --id 1
Resource operatingsystem not found by id '1'
```
For some reason default OS is not present there anymore.
Basically, that's why me & Alex are always using some complex logic with default entities like  'find the default entity by name, not id, if it was found - then use it, otherwise create a new entity instead' and encouraging others not to rely on id=1.
Default org/loc are less likely to be affected, but still, who knows, maybe next time after roll back their ids will be 3 and 4 correspondingly.

**Comment by rplevka** [comment]
@abalakh @svtkachenko works for me guys, I just wanted to cast some light into the default ids for the taxonomies and capsules :)

**Comment by rplevka** [comment]
perhaps just a nitpick if we want to really minimize the risks - I would search on url with the hostname of the server and the port of 9090 instead of the name.

**Comment by svtkachenko** [comment]
Update: now searching by url and not id.
```python
% nosetests tests/foreman/cli/test_host.py:{HostCreateTestCase,HostDeleteTestCase,HostUpdateTestCase,HostParameterTestCase} tests/foreman/cli/test_hostgroup.py:HostGroupTestCase.test_positive_create_with_puppet_ca_proxy tests/foreman/api/test_host.py:HostTestCase.{test_positive_create_with_puppet_proxy,test_positive_create_with_puppet_ca_proxy,test_positive_update_puppet_proxy,test_positive_update_puppet_ca_proxy} tests/foreman/api/test_hostgroup.py:HostGroupTestCase.{test_positive_create_with_puppet_ca_proxy,test_positive_create_with_realm,test_positive_create_with_puppet_proxy,test_positive_create_with_content_source,test_positive_update_puppet_ca_proxy,test_positive_update_realm,test_positive_update_puppet_proxy,test_positive_update_content_source}      
....................................S..SSSS
----------------------------------------------------------------------
Ran 43 tests in 1748.194s
OK (SKIP=5)
```

**Comment by rplevka** [comment]
works for me,
ACK

File: tests/foreman/api/test_organization.py [hunk 1]
```diff
@@ -349,9 +349,9 @@ def test_positive_add_smart_proxy(self):
         @CaseLevel: Integration
         """
         # Every Satellite has a built-in smart proxy, so let's find it
-        smart_proxies = entities.SmartProxy().search()
-        self.assertGreater(len(smart_proxies), 0)
```
Comment (oshtaier) [inline tests/foreman/api/test_organization.py]: Is there any reason to remove this verification? Does it mean that you absolutely sure that default capsule exist?

File: tests/foreman/api/test_smartproxy.py [hunk 2]
```diff
@@ -41,9 +42,10 @@ def setUpClass(cls):
         existing smart proxy should always succeed.
         """
         super(SmartProxyMissingAttrTestCase, cls).setUpClass()
-        smart_proxies = entities.SmartProxy().search()
-        assert len(smart_proxies) > 0
```
Comment (oshtaier) [inline tests/foreman/api/test_smartproxy.py]: Same here + in case that assertion failed we will not run tests at all, so have error in the code, not failed cases

File: tests/foreman/cli/test_host.py [hunk 3]
```diff
@@ -383,9 +383,9 @@ class HostUpdateTestCase(CLITestCase):
     def setUp(self):
         """Create a host to reuse later"""
         super(HostUpdateTestCase, self).setUp()
-        self.proxies = Proxy.list()
-        self.assertGreater(len(self.proxies), 0)
```
Comment (oshtaier) [inline tests/foreman/cli/test_host.py]: Same question here

**Comment by oshtaier** [comment]
ACK pending questions. Feel free to merge if I not respond in one day on your answers

**Comment by sghai** [comment]
@svtkachenko ping ?

File: tests/foreman/api/test_organization.py (hunk 1 above)
Comment (abalakh) [inline tests/foreman/api/test_organization.py]: If default capsule doesn't exist, then search will return empty list and `.search()[0]` will lead to `'list index out of range` exception, so we won't miss the failure here. For me 'list index out of range` is pretty clear type of error, but it may be not that clear to everyone - asserting len of search results imo leaves no questions of why test failed, so maybe we should keep this assertion.

File: tests/foreman/api/test_smartproxy.py (hunk 2 above)
Comment (abalakh) [inline tests/foreman/api/test_smartproxy.py]: If i'm not wrong, the only way to mark test as failed is to fail unittest's assertion. Default python assertion will mark class as errored and we can't use unittest here, so class will be marked as 'error' regardless. The only question here is whether  `list index out of range` message is as clear as `0 != 1` message.

File: tests/foreman/cli/test_gpgkey.py [hunk 4]
```diff
@@ -135,9 +135,7 @@ def test_positive_create_with_default_org(self):
 
         @assert: gpg key is created
         """
-        result = Org.list()
-        self.assertGreater(len(result), 0, 'No organization found')
```
Comment (abalakh) [inline tests/foreman/cli/test_gpgkey.py]: `.list()` may return empty list if command succeeded but stdout is empty (no results found), so assertion after list command is a good thing, `.info()` will have non-zero return-code in case of entity not found, and as we catch return-codes of all the executed commands, if org wasn't found we'll raise an exception here.
This change is good as for me as both ways we won't miss failure but this new way we are working with default org directly and not assuming that the first entity returned by `.list()` command is the default org (which is correct right now but no guarranties it won't be changed in future - for some commands the list is sorted alphabetically by name, not id).

File: tests/foreman/api/test_organization.py (hunk 1 above)
Comment (abalakh) [inline tests/foreman/api/test_organization.py]: Another benefit of using assertion here is the fact it will result in test marked with `fail`, not `error`. In some meaning it's debatable too - in case we've lost default capsule it's not test's fault imo, but let's better stick to failures instead of errors.

**Comment by svtkachenko** [comment]
Asserts restored according to @oshtaier and @abalakh comments.
Also added skip decorators because of found bugs.

File: tests/foreman/api/test_smartproxy.py (hunk 2 above)
Comment (rplevka) [inline tests/foreman/api/test_smartproxy.py]: +1 to what @abalakh said. Perhaps, just to make the assertion failure more descriptive, we could add a failure message, such as "No Capsules retrieved via API"

File: tests/foreman/api/test_organization.py (hunk 1 above)
Comment (rplevka) [inline tests/foreman/api/test_organization.py]: @abalakh ,
I've actually started to fight with improper assertions here https://github.com/SatelliteQE/robottelo/issues/4041
:)

**Comment by oshtaier** [comment]
ACK pending last @abalakh and @rplevka comments

File: tests/foreman/cli/test_hostgroup.py [hunk 5]
```diff
@@ -316,21 +316,28 @@ def test_positive_create_with_multiple_entities(self):
...
             'medium-id': media['id'],
             'operatingsystem-id': os['id'],
-        })
+        }
+        # If bug is open provide LCE id as parameter
+        # because LCE name cause errors
+        if bz_bug_is_open(1395254):
+            make_hostgroup_params['lifecycle-environment-id'] = lce['id']
+        else:
+            make_hostgroup_params['lifecycle-environment'] = lce['name']
```
Comment (abalakh) [inline tests/foreman/cli/test_hostgroup.py]: nitpick: i'd rather return back `lce['name']` to the list of parameters, and if bug is opened i'd pop it & used `id` instead. The reason is simple - you'll design the test in the way it supposed to work and the workaround will be applied only in case bug is open, right now workaround for params is applied regardless of bug state :) And also it adds readability as it'll become clear which params are supposed to be passed to `make_hostgroup()`
